sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1248 @@
+from ..utils.extmath import safe_sparse_dot
+from ..utils.validation import check_is_fitted
+from ..utils.multiclass import _check_partial_fit_first_call
+
+
+_STOCHASTIC_ALGOS = ['sgd', 'adam']
+
+
+def _pack(coefs_, intercepts_):
+    """Pack the parameters into a single vector."""
+    return np.hstack([l.ravel() for l in coefs_ + intercepts_])
+
+
+class BaseMultilayerPerceptron(six.with_metaclass(ABCMeta, BaseEstimator)):
+    """Base class for MLP classification and regression.
+
+    Warning: This class should not be used directly.
+    Use derived classes instead.
+    """
+
+    @abstractmethod
+    def __init__(self, hidden_layer_sizes, activation, algorithm,
+                 alpha, batch_size, learning_rate, learning_rate_init, power_t,
+                 max_iter, loss, shuffle, random_state, tol, verbose,
+                 warm_start, momentum, nesterovs_momentum, early_stopping,
+                 validation_fraction, beta_1, beta_2, epsilon):
+        self.activation = activation
+        self.algorithm = algorithm
+        self.alpha = alpha
+        self.batch_size = batch_size
+        self.learning_rate = learning_rate
+        self.learning_rate_init = learning_rate_init
+        self.power_t = power_t
+        self.max_iter = max_iter
+        self.loss = loss
+        self.hidden_layer_sizes = hidden_layer_sizes
+        self.shuffle = shuffle
+        self.random_state = random_state
+        self.tol = tol
+        self.verbose = verbose
+        self.warm_start = warm_start
+        self.momentum = momentum
+        self.nesterovs_momentum = nesterovs_momentum
+        self.early_stopping = early_stopping
+        self.validation_fraction = validation_fraction
+        self.beta_1 = beta_1
+        self.beta_2 = beta_2
+        self.epsilon = epsilon
+
+    def _unpack(self, packed_parameters):
+        """Extract the coefficients and intercepts from packed_parameters."""
+        for i in range(self.n_layers_ - 1):
+            start, end, shape = self._coef_indptr[i]
+            self.coefs_[i] = np.reshape(packed_parameters[start:end], shape)
+
+            start, end = self._intercept_indptr[i]
+            self.intercepts_[i] = packed_parameters[start:end]
+
+    def _forward_pass(self, activations, with_output_activation=True):
+        """Perform a forward pass on the network by computing the values
+        of the neurons in the hidden layers and the output layer.
+
+        Parameters
+        ----------
+        activations: list, length = n_layers - 1
+            The ith index of the list holds the values of the ith layer.
+
+        with_output_activation : bool, default True
+            If True, the output passes through the output activation
+            function, which is either the softmax function or the
+            logistic function
+        """
+        # Iterate over the hidden layers
+        for i in range(self.n_layers_ - 1):
+            activations[i + 1] = safe_sparse_dot(activations[i],
```

Comment:
Do we really need sparse dot? The intermediate layers cannot be sparse, right? Only the input